modules/delegates.js

```diff
@@ -353,112 +353,146 @@ __private.decryptPassphrase = function(encryptedPassphrase, password) {
-							) {
-								return setImmediate(
-									cb,
-									`Failed to add vote, delegate "${
-										account.username
-									}" already voted for`
-								);
-							}
-
-							if (
-								math === '-' &&
-								(delegates === null || delegates.indexOf(publicKey) === -1)
-							) {
-								return setImmediate(
-									cb,
-									`Failed to remove vote, delegate "${
-										account.username
-									}" was not voted for`
-								);
-							}
-
-							return setImmediate(cb);
-						},
-						tx
-					);
-				},
-				err => {
-					if (err) {
-						return setImmediate(cb, err);
-					}
-
-					const totalVotes = existingVotes + additions - removals;
+			voteAddress = modules.accounts.generateAddressByPublicKey(votePublicKey);
+		} catch (e) {
+			library.logger.error(e.stack);
+			return setImmediate(cb, e);
+		}
+		return {
+			action,
+			address: voteAddress,
+		};
+	});
 
-					if (totalVotes > constants.activeDelegates) {
-						const exceeded = totalVotes - constants.activeDelegates;
+	async.waterfall(
+		[
+			function getVotedAccounts(waterfallCb) {
+				modules.accounts.getAccount(
+					{ publicKey: senderPublicKey },
+					(err, account) => {
+						if (err) {
+							return setImmediate(waterfallCb, err);
+						}
+
+						if (!account) {
+							return setImmediate(waterfallCb, 'Account not found');
+						}
+
+						const delegates =
+							state === 'confirmed' ? account.delegates : account.u_delegates;
+						const votedAddresses = Array.isArray(delegates)
+							? delegates.map(delegate =>
+									modules.accounts.generateAddressByPublicKey(delegate)
+								)
+							: [];
+
+						return setImmediate(waterfallCb, null, votedAddresses);
+					},
+					tx
+				);
+			},
+			function validateVotes(existingVotedAddresses, waterfallCb) {
+				modules.accounts.getAccounts(
+					{
+						address: actionWithAddresses.map(({ address }) => address),
+						isDelegate: 1,
+					},
+					(err, votesAccounts) => {
+						if (err) {
+							return setImmediate(waterfallCb, err);
+						}
+
+						if (
+							!votesAccounts ||
+							votesAccounts.length < actionWithAddresses.length
+						) {
+							return setImmediate(waterfallCb, 'Delegate not found');
+						}
+
+						const upvoteAccounts = votesAccounts.filter(voteAccount =>
+							actionWithAddresses.find(
+								actionWithAddress =>
+									actionWithAddress.action === '+' &&
+									actionWithAddress.address === voteAccount.address
+							)
+						);
+						const downvoteAccounts = votesAccounts.filter(voteAccount =>
+							actionWithAddresses.find(
+								actionWithAddress =>
+									actionWithAddress.action === '-' &&
+									actionWithAddress.address === voteAccount.address
+							)
+						);
 
-						return setImmediate(
-							cb,
-							`Maximum number of ${
-								constants.activeDelegates
-							} votes exceeded (${exceeded} too many)`
+						const invalidUpvoteAccount = upvoteAccounts.find(
+							voteAccount =>
```

Comment:
review logic (performance)

Reply:
There can be one performance improvement. But I think it'll make the code more complicated. I think performance benefit is not so much.